File: pulse-kotlin/src/commonTest/kotlin/kotlin/com/fluxchess/pulse/kotlin/SquareTest.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package kotlin.com.fluxchess.pulse.kotlin

import com.fluxchess.pulse.kotlin.fileOf
import com.fluxchess.pulse.kotlin.files
import com.fluxchess.pulse.kotlin.rankOf
import com.fluxchess.pulse.kotlin.ranks
import com.fluxchess.pulse.kotlin.squareOf
import com.fluxchess.pulse.kotlin.squares
import kotlin.test.Test
import kotlin.test.assertEquals

class SquareTest {
	@Test
	fun `Square array should be valid`() {
		for (rank in ranks.indices) {
			for (file in files.indices) {
				assertEquals(squares[rank * ranks.size + file], squareOf(file, rank))
			}
		}
	}

	@Test
	fun `When creating a square it should save and return file and rank correctly`() {
		ranks.forEach { rank ->
			files.forEach { file ->
				val square = squareOf(file, rank)
				assertEquals(file, fileOf(square))
				assertEquals(rank, rankOf(square))
			}
		}
	}
}

File: pulse-go/internal/pulse/square_test.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package pulse

import (
	"testing"
)

func TestSquare(t *testing.T) {
	t.Run("Square array should be valid", func(t *testing.T) {
		for rank := range Ranks {
			for file := range Files {
				index := rank*len(Ranks) + file
				square := SquareOf(file, rank)
				if Squares[index] != square {
					t.Errorf("wanted squares at index %d to be %v, got %v", index, square, Squares[index])
				}
			}
		}
	})

	t.Run("When creating a square it should save and return file and rank correctly", func(t *testing.T) {
		for _, rank := range Ranks {
			for _, file := range Files {
				square := SquareOf(file, rank)
				if FileOf(square) != file {
					t.Errorf("wanted file of square to be %v, got %v", file, FileOf(square))
				}
				if RankOf(square) != rank {
					t.Errorf("wanted rank of square to be %v, got %v", rank, RankOf(square))
				}
			}
		}
	})
}
